File: packages/typescript/tests/edit.test.ts
import { describe, it, expect } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { openExisting } from '../src/edit';
import { create, SkpWriteError, Point3 } from '../src/create';
import { parseSkp } from '../src/index';

/**
 * Tests for edit.ts - loading an existing legacy .skp file and rebuilding
 * it as a new SkpBuilder, ported from Python's test_edit.py (see edit.ts's
 * own docstring for the exact scope and known fidelity gaps this suite
 * exercises).
 */

const FIXTURES = path.join(__dirname, 'fixtures');
const SQUARE: Point3[] = [
  [0.0, 0.0, 0.0],
  [100.0, 0.0, 0.0],
  [100.0, 100.0, 0.0],
  [0.0, 100.0, 0.0],
];

function toBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
}

describe('openExisting', () => {
  it('rejects a VFF (modern) source', () => {
    expect(() => openExisting(path.join(FIXTURES, 'SU_File.skp'))).toThrow(/not a legacy-format/);
  });

  it('rejects a nonexistent file', () => {
    expect(() => openExisting(path.join(FIXTURES, 'does_not_exist.skp'))).toThrow();
  });

  it('round-trips a simple file (materials, layers, a definition, 2 instances, a root face)', () => {
    const builder = create();
    const red = builder.addMaterial('Red', [255, 0, 0]);
    const roof = builder.addLayer('Roof');
    const chair = builder.addComponentDefinition('Chair', (def) => {
      def.addFace([[0, 0, 0], [20, 0, 0], [20, 20, 0], [0, 20, 0]], { material: red });
    });
    builder.addInstance(chair, { translation: [0, 0, 0] });
    builder.addInstance(chair, { translation: [50, 0, 0] });
    builder.addFace(SQUARE, { material: red, layer: roof });

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    expect(rebuilt.root.faces.length).toBe(1);
    expect(rebuilt.root.instances.length).toBe(2);
    expect(rebuilt.definitions.size).toBe(1);
    const chairDefn = [...rebuilt.definitions.values()][0];
    expect(chairDefn.name).toBe('Chair');
    expect(chairDefn.faces.length).toBe(1);
    expect(rebuilt.materials.map((m) => m.name)).toEqual(['Red']);
    expect(rebuilt.layers.map((l) => l.name)).toContain('Roof');
  });

  it('preserves instance translation', () => {
    const builder = create();
    const post = builder.addComponentDefinition('Post', (def) => {
      def.addFace([[0, 0, 0], [5, 0, 0], [5, 5, 0], [0, 5, 0]]);
    });
    builder.addInstance(post, { translation: [37.5, -12.25, 8.0] });

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    const inst = rebuilt.root.instances[0];
    expect(inst.matrix[9]).toBeCloseTo(37.5, 9);
    expect(inst.matrix[10]).toBeCloseTo(-12.25, 9);
    expect(inst.matrix[11]).toBeCloseTo(8.0, 9);
  });

  it('preserves instance hidden state and layer color', () => {
    const builder = create();
    const roof = builder.addLayer('Roof', { color: [150, 75, 30], hidden: true });
    const post = builder.addComponentDefinition('Post', (def) => {
      def.addFace([[0, 0, 0], [5, 0, 0], [5, 5, 0], [0, 5, 0]]);
    });
    builder.addInstance(post, { hidden: true, layer: roof });

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    expect(rebuilt.root.instances[0].hidden).toBe(true);
    const roofLayer = rebuilt.layers.find((l) => l.name === 'Roof')!;
    expect(roofLayer.color).toEqual({ r: 150, g: 75, b: 30 });
    expect(roofLayer.hidden).toBe(true);
  });

  it('replays nested definitions in dependency order (Wheel before Car)', () => {
    const builder = create();
    const wheel = builder.addComponentDefinition('Wheel', (def) => {
      def.addFace([[0, 0, 0], [10, 0, 0], [10, 10, 0], [0, 10, 0]]);
    });
    const car = builder.addComponentDefinition('Car', (def) => {
      def.addFace([[0, 0, 0], [100, 0, 0], [100, 50, 0], [0, 50, 0]]);
      def.addInstance(wheel, { translation: [10, 10, 0] });
      def.addInstance(wheel, { translation: [80, 10, 0] });
    });
    builder.addInstance(car);

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    const byName = new Map([...rebuilt.definitions.values()].map((d) => [d.name, d]));
    expect(new Set(byName.keys())).toEqual(new Set(['Wheel', 'Car']));
    expect(byName.get('Car')!.instances.length).toBe(2);
  });

  it('materials and layers are reusable (not duplicated) after replay', () => {
    const builder = create();
    const red = builder.addMaterial('Red', [255, 0, 0]);
    const roof = builder.addLayer('Roof');
    builder.addFace(SQUARE, { material: red, layer: roof });

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    expect(newBuilder.materialsByName.has('Red')).toBe(true);
    expect(newBuilder.layersByName.has('Roof')).toBe(true);
    newBuilder.addFace(
      [[300, 0, 0], [310, 0, 0], [310, 10, 0], [300, 10, 0]],
      { material: newBuilder.materialsByName.get('Red'), layer: newBuilder.layersByName.get('Roof') }
    );
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));
    expect(rebuilt.materials.length).toBe(1); // still just "Red" - reused, not duplicated
    expect(rebuilt.root.faces.length).toBe(2);
  });

  it('returns definitions by name, directly usable for a new placement', () => {
    const builder = create();
    const wheel = builder.addComponentDefinition('Wheel', (def) => {
      def.addFace([[0, 0, 0], [10, 0, 0], [10, 10, 0], [0, 10, 0]]);
    });
    builder.addInstance(wheel, { translation: [0, 0, 0] });

    const { builder: newBuilder, definitions } = openExisting(toBuffer(builder.toBytes()));
    expect(definitions.has('Wheel')).toBe(true);
    newBuilder.addInstance(definitions.get('Wheel')!, { translation: [100, 0, 0] });
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));
    expect(rebuilt.definitions.size).toBe(1); // still just one Wheel definition
    expect(rebuilt.root.instances.length).toBe(2); // but now placed twice
  });

  it('a genuinely new material/layer/definition/group is rejected after replay', () => {
    const builder = create();
    builder.addFace(SQUARE);

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    expect(() => newBuilder.addMaterial('Chrome', [180, 180, 185])).toThrow(SkpWriteError);
    expect(() => newBuilder.addLayer('Extra')).toThrow(SkpWriteError);
    expect(() => newBuilder.addComponentDefinition('New', () => {})).toThrow(SkpWriteError);
    expect(() => newBuilder.addGroup(() => {}, { name: 'NewGroup' })).toThrow(SkpWriteError);
  });

  it('a positioned texture round-trips', () => {
    // Minimal 4x4 solid-color PNG (same generator as create.test.ts's
    // Textures suite - kept local here to avoid a cross-file test dependency).
    function makeTestPng(): Uint8Array {
      function crc32(buf: number[]): number {
        const table: number[] = [];
        for (let n = 0; n < 256; n++) {
          let c = n;
          for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
          table[n] = c >>> 0;
        }
        let crc = 0xffffffff;
        for (const b of buf) crc = table[(crc ^ b) & 0xff] ^ (crc >>> 8);
        return (crc ^ 0xffffffff) >>> 0;
      }
      function u32be(v: number): number[] {
        return [(v >>> 24) & 0xff, (v >>> 16) & 0xff, (v >>> 8) & 0xff, v & 0xff];
      }
      function chunk(tag: number[], data: number[]): number[] {
        return [...u32be(data.length), ...tag, ...data, ...u32be(crc32([...tag, ...data]))];
      }
      const size = 4;
      const rgb = [200, 50, 50];
      const rawRows: number[][] = [];
      for (let y = 0; y < size; y++) {
        const row = [0];
        for (let x = 0; x < size; x++) row.push(...rgb);
        rawRows.push(row);
      }
      const rawData = rawRows.flat();
      let adler1 = 1;
      let adler2 = 0;
      for (const b of rawData) {
        adler1 = (adler1 + b) % 65521;
        adler2 = (adler2 + adler1) % 65521;
      }
      const deflate: number[] = [];
      deflate.push(1, rawData.length & 0xff, (rawData.length >> 8) & 0xff, ~rawData.length & 0xff, (~rawData.length >> 8) & 0xff, ...rawData);
      const zlibStream = [0x78, 0x01, ...deflate, (adler2 >> 8) & 0xff, adler2 & 0xff, (adler1 >> 8) & 0xff, adler1 & 0xff];
      const sig = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];
      const ihdr = chunk([0x49, 0x48, 0x44, 0x52], [...u32be(size), ...u32be(size), 8, 2, 0, 0, 0]);
      const idat = chunk([0x49, 0x44, 0x41, 0x54], zlibStream);
      const iend = chunk([0x49, 0x45, 0x4e, 0x44], []);
      return Uint8Array.from([...sig, ...ihdr, ...idat, ...iend]);
    }

    const builder = create();
    const png = makeTestPng();
    const brick = builder.addTextureMaterial('Brick', png, 'brick.png');
    builder.addFace(SQUARE, {
      material: brick,
      frontUv: [
        [[0.0, 0.0, 0.0], [0.0, 0.0]],
        [[50.0, 0.0, 0.0], [1.0, 0.0]],
        [[0.0, 50.0, 0.0], [0.0, 1.0]],
      ],
    });

    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    const face = rebuilt.root.faces[0];
    expect(face.uvTransform).not.toBeNull();
    expect(rebuilt.materials[0].texture).not.toBeNull();
    expect(Array.from(rebuilt.materials[0].texture!.data as Uint8Array)).toEqual(Array.from(png));
  });

  it('a face with a hole round-trips without a "hole" warning', () => {
    const wall: Point3[] = [[0, 0, 0], [200, 0, 0], [200, 100, 0], [0, 100, 0]];
    const window: Point3[] = [[80, 30, 0], [120, 30, 0], [120, 70, 0], [80, 70, 0]];
    const builder = create();
    builder.addFace(wall, { holes: [window] });

    const { builder: newBuilder, warnings } = openExisting(toBuffer(builder.toBytes()));
    expect(warnings.some((w) => w.includes('hole'))).toBe(false);
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));
    expect(rebuilt.root.faces[0].loops.length).toBe(2);
  });

  it('an empty source can still be saved after adding geometry', () => {
    const { builder: newBuilder } = openExisting(path.join(FIXTURES, 'blank_v17.skp'));
    expect(() => newBuilder.toBytes()).toThrow(/no geometry/);
    newBuilder.addFace(SQUARE);
    expect(newBuilder.toBytes().length).toBeGreaterThan(0);
  });

  it('per-edge flags collapse to a per-face approximation (documented gap)', () => {
    const builder = create();
    builder.addFace(SQUARE, { hiddenEdges: true, softEdges: true, smoothEdges: true });
    const { builder: newBuilder } = openExisting(toBuffer(builder.toBytes()));
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));
    expect(rebuilt.root.edges.every((e) => e.hidden && e.soft && e.smooth)).toBe(true);
  });
});

describe('Real-world fixtures (non-writer-authored)', () => {
  // The true stress test for edit.ts, since every test above only
  // exercises content this package's own writer already produces (a much
  // narrower subset of what real SketchUp files contain).
  const fixtures = ['capilla_quiroz_v17.skp', 'gondola_v20.skp'];

  for (const fixtureName of fixtures) {
    it(`round-trips ${fixtureName} without crashing`, () => {
      const fixturePath = path.join(FIXTURES, fixtureName);
      if (!fs.existsSync(fixturePath)) return; // skip if not present, same as Python's suite
      const { builder: newBuilder } = openExisting(fixturePath);
      const data = newBuilder.toBytes();
      // Self-parses without throwing - the authoritative "is this
      // structurally valid" check for a file this large/real.
      expect(() => parseSkp(toBuffer(data))).not.toThrow();
    }, 30000);
  }

  it('capilla_quiroz_v17.skp preserves almost all geometry', () => {
    const fixturePath = path.join(FIXTURES, 'capilla_quiroz_v17.skp');
    if (!fs.existsSync(fixturePath)) return;
    const origBuf = fs.readFileSync(fixturePath);
    const orig = parseSkp(toBuffer(new Uint8Array(origBuf.buffer, origBuf.byteOffset, origBuf.byteLength)));
    const { builder: newBuilder } = openExisting(fixturePath);
    const rebuilt = parseSkp(toBuffer(newBuilder.toBytes()));

    let origTotal = orig.root.faces.length;
    for (const d of orig.definitions.values()) origTotal += d.faces.length;
    let rebuiltTotal = rebuilt.root.faces.length;
    for (const d of rebuilt.definitions.values()) rebuiltTotal += d.faces.length;

    // At most a handful of faces (e.g. a degenerate UV correspondence) are
    // expected to be skipped, never a large fraction - a big drop would
    // indicate silent corruption, not a legitimately-scoped gap.
    expect(rebuiltTotal).toBeGreaterThanOrEqual(origTotal - 5);
    expect(rebuilt.root.instances.length).toBe(orig.root.instances.length);
    expect(rebuilt.definitions.size).toBe(orig.definitions.size);
  }, 30000);
});

File: packages/cpp/tests/edit_test.cpp
// Tests for openskp::open_existing (packages/cpp/src/edit.cpp), the C++ port of
// packages/python/src/openskp/edit.py. Mirrors packages/python/tests/test_edit.py's approach:
// build a file with SkpBuilder, save it, reopen it with open_existing(), and check the rebuilt
// builder's own re-parsed output preserves the original's structure - plus an end-to-end pass
// against a real SketchUp-SDK-authored fixture (capilla_quiroz_v17.skp) already used elsewhere in
// this test suite.

#include <gtest/gtest.h>

#include <openskp/openskp.hpp>

#include "test_helpers.hpp"

namespace openskp {
namespace {

std::filesystem::path temp_skp(const char* stem) {
  return std::filesystem::temp_directory_path() /
         (std::string("openskp_edit_test_") + stem + ".skp");
}

// A syntactically-tiny PNG - just the magic bytes plus filler, matching create_test.cpp's own
// fake_png_bytes() (not shared across translation units, so duplicated here).
ByteBuffer fake_png_bytes() {
  ByteBuffer data = {0x89, 'P', 'N', 'G', '\r', '\n', 0x1a, '\n'};
  data.insert(data.end(), 64, 0x42);
  return data;
}

TEST(Edit, RejectsAModernNonLegacyFile) {
  EXPECT_THROW(open_existing(test::fixture("Untitled.skp")), SkpWriteError);
}

TEST(Edit, RejectsAMissingFile) {
  EXPECT_THROW(open_existing(test::fixture("does_not_exist_at_all.skp")), SkpWriteError);
}

TEST(Edit, RoundTripsAFreshlyBuiltFile) {
  auto path = temp_skp("roundtrip");
  {
    auto builder = create();
    int red = builder->add_material("Red", Color3{255, 0, 0});
    int roof = builder->add_layer("Roof");

    auto& chair = builder->add_component_definition("Chair");
    chair.add_face({{0, 0, 0}, {20, 0, 0}, {20, 20, 0}, {0, 20, 0}});
    chair.close();
    InstanceOptions iopts;
    iopts.translation = {200, 0, 0};
    builder->add_instance(chair, iopts);

    FaceOptions opts;
    opts.material = red;
    opts.layer = roof;
    builder->add_face({{0, 0, 0}, {100, 0, 0}, {100, 100, 0}, {0, 100, 0}}, opts);

    builder->save(path);
  }

  OpenExistingResult result = open_existing(path);
  ASSERT_NE(result.builder, nullptr);
  EXPECT_EQ(result.builder->materials_by_name.count("Red"), 1u);
  EXPECT_EQ(result.builder->layers_by_name.count("Roof"), 1u);
  ASSERT_EQ(result.definitions.count("Chair"), 1u);

  // The rebuilt file should carry the same face/material/layer/instance/definition counts as the
  // original when parsed back with our own reader.
  ByteBuffer rebuilt_bytes = result.builder->to_bytes();
  SkpModel rebuilt = SkpFile::from_buffer(rebuilt_bytes).parse();
  EXPECT_EQ(rebuilt.materials.size(), 1u);
  EXPECT_EQ(rebuilt.layers.size(), 2u);  // Layer0 + Roof
  EXPECT_EQ(rebuilt.definitions.size(), 1u);
  EXPECT_EQ(rebuilt.root().faces.size(), 1u);
  EXPECT_EQ(rebuilt.root().instances.size(), 1u);

  std::filesystem::remove(path);
}

TEST(Edit, DefaultProjectedTextureGetsAnExplicitUvPinNotLeftCorrupted) {
  // Found via cross-language analysis (2026-08-28): replay_materials wrote every rebuilt texture
  // material with the library's (until now) corrupted default applied height, and replay_uv only
  // replayed a face's UV when it already had an explicit uv_transform - leaving a DEFAULT-projected
  // textured face's material (and thus its real SketchUp rendering) corrupted end to end. Both are
  // fixed; this checks the rebuilt face actually gets a pin (not just that some texture data
  // round-tripped) and that the material's applied height isn't the corrupted sentinel.
  auto png_path = std::filesystem::temp_directory_path() / "openskp_edit_test_uv_texture.png";
  {
    std::ofstream f(png_path, std::ios::binary);
    ByteBuffer png = fake_png_bytes();
    f.write(reinterpret_cast<const char*>(png.data()), static_cast<std::streamsize>(png.size()));
  }
  auto path = temp_skp("default_projected_texture");
  {
    auto builder = create();
    int brick = builder->add_texture_material("Brick", png_path);
    FaceOptions opts;
    opts.material = brick;
    builder->add_face({{0, 0, 0}, {100, 0, 0}, {100, 100, 0}, {0, 100, 0}}, opts);  // no front_uv
    builder->save(path);
  }
  std::filesystem::remove(png_path);

  OpenExistingResult result = open_existing(path);
  ASSERT_NE(result.builder, nullptr);
  ByteBuffer rebuilt_bytes = result.builder->to_bytes();
  SkpModel rebuilt = SkpFile::from_buffer(rebuilt_bytes).parse();
  std::filesystem::remove(path);

  ASSERT_EQ(rebuilt.materials.size(), 1u);
  ASSERT_TRUE(rebuilt.materials[0].texture.has_value());
  EXPECT_DOUBLE_EQ(rebuilt.materials[0].texture->width, 1.0);
  EXPECT_DOUBLE_EQ(rebuilt.materials[0].texture->height, 1.0);
  ASSERT_EQ(rebuilt.root().faces.size(), 1u);
  EXPECT_TRUE(rebuilt.root().faces.begin()->second.uv_transform.has_value());
}

TEST(Edit, GenuinelyEmptyInstanceNameIsPreservedNotReplaced) {
  // Found via cross-language analysis (2026-08-28): add_instance's own name.value_or(definition
  // name) fallback and replay_instance's `if (!inst.name.empty()) options.name = inst.name;` both
  // silently replaced a genuinely empty instance name with its definition's name - a real
  // difference, not cosmetic (a later rename of the definition would no longer show through).
  auto path = temp_skp("empty_instance_name");
  {
    auto builder = create();
    auto& box = builder->add_component_definition("Box");
    box.add_face({{0, 0, 0}, {10, 0, 0}, {10, 10, 0}, {0, 10, 0}});
    box.close();
    InstanceOptions iopts;
    iopts.name = "";
    builder->add_instance(box, iopts);
    builder->save(path);
  }

  SkpModel source = SkpFile::open(path).parse();
  ASSERT_EQ(source.root().instances.size(), 1u);
  EXPECT_EQ(source.root().instances[0].name, "");

  OpenExistingResult result = open_existing(path);
  ASSERT_NE(result.builder, nullptr);
  ByteBuffer rebuilt_bytes = result.builder->to_bytes();
  SkpModel rebuilt = SkpFile::from_buffer(rebuilt_bytes).parse();
  std::filesystem::remove(path);

  ASSERT_EQ(rebuilt.root().instances.size(), 1u);
  EXPECT_EQ(rebuilt.root().instances[0].name, "");
}

TEST(Edit, GenuinelyEmptyDefinitionNameIsPreservedNotReplaced) {
  // Found via cross-language analysis (2026-08-28), same bug class as the empty instance name
  // case above: `defn.name.empty() ? ("Definition" + std::to_string(def_id)) : defn.name` silently
  // replaced a genuinely empty definition name with a fabricated one. SketchUp Groups are
  // internally just unnamed component definitions (unlike Components, which SketchUp
  // auto-names), so an empty name is common in real files.
  auto path = temp_skp("empty_definition_name");
  {
    auto builder = create();
    auto& box = builder->add_component_definition("");
    box.add_face({{0, 0, 0}, {10, 0, 0}, {10, 10, 0}, {0, 10, 0}});
    box.close();
    builder->add_instance(box, {});
    builder->save(path);
  }

  SkpModel source = SkpFile::open(path).parse();
  ASSERT_EQ(source.definitions.size(), 1u);
  EXPECT_EQ(source.definitions.begin()->second.name, "");

  OpenExistingResult result = open_existing(path);
  ASSERT_NE(result.builder, nullptr);
  ByteBuffer rebuilt_bytes = result.builder->to_bytes();
  SkpModel rebuilt = SkpFile::from_buffer(rebuilt_bytes).parse();
  std::filesystem::remove(path);

  ASSERT_EQ(rebuilt.definitions.size(), 1u);
  EXPECT_EQ(rebuilt.definitions.begin()->second.name, "");
}

TEST(Edit, ReturnedBuilderCanAddMoreGeometryAndCannotRegisterNewSections) {
  auto path = temp_skp("addmore");
  {
    auto builder = create();
    builder->add_face({{0, 0, 0}, {10, 0, 0}, {10, 10, 0}, {0, 10, 0}});
    builder->save(path);
  }

  OpenExistingResult result = open_existing(path);
  // More geometry is fine.
  EXPECT_NO_THROW(result.builder->add_face({{50, 50, 0}, {60, 50, 0}, {60, 60, 0}, {50, 60, 0}}));
  // Registering a genuinely new material/layer/definition is not - the file format's ordering
  // requirement (materials/layers/definitions before any geometry) was already satisfied by
  // replaying the source's own root-level geometry.
  EXPECT_THROW(result.builder->add_material("New", Color3{1, 2, 3}), SkpWriteError);
  EXPECT_THROW(result.builder->add_layer("New"), SkpWriteError);

  ByteBuffer bytes = result.builder->to_bytes();
  SkpModel model = SkpFile::from_buffer(bytes).parse();
  EXPECT_EQ(model.root().faces.size(), 2u);

  std::filesystem::remove(path);
}

TEST(Edit, DefinitionsMapAllowsPlacingMoreInstancesOfAnExistingDefinition) {
  auto path = temp_skp("moreinstances");
  {
    auto builder = create();
    auto& wheel = builder->add_component_definition("Wheel");
    wheel.add_face({{0, 0, 0}, {10, 0, 0}, {10, 10, 0}, {0, 10, 0}});
    wheel.close();
    builder->add_instance(wheel);
    builder->save(path);
  }

  OpenExistingResult result = open_existing(path);
  ASSERT_EQ(result.definitions.count("Wheel"), 1u);
  InstanceOptions opts;
  opts.translation = {500, 0, 0};
  EXPECT_NO_THROW(result.builder->add_instance(*result.definitions.at("Wheel"), opts));

  ByteBuffer bytes = result.builder->to_bytes();
  SkpModel model = SkpFile::from_buffer(bytes).parse();
  EXPECT_EQ(model.root().instances.size(), 2u);

  std::filesystem::remove(path);
}

TEST(Edit, EmptyBlankScaffoldHasNoGeometryToReplayButStillOpens) {
  // blank_v17.skp has 0 materials/layers-beyond-Layer0/definitions/root entities - exercises the
  // "nothing to replay" path (no crash, an empty-but-valid builder) distinctly from every other
  // test here, which all replay at least one face.
  OpenExistingResult result = open_existing(test::fixture("blank_v17.skp"));
  ASSERT_NE(result.builder, nullptr);
  EXPECT_TRUE(result.definitions.empty());
  // Can't call to_bytes() yet - matches create()'s own "no geometry added" guard - but adding a
  // face and saving should work.
  EXPECT_THROW(result.builder->to_bytes(), SkpWriteError);
  EXPECT_NO_THROW(result.builder->add_face({{0, 0, 0}, {1, 0, 0}, {1, 1, 0}}));
  EXPECT_NO_THROW(result.builder->to_bytes());
}

// ---------------------------------------------------------------------------------------------
// End-to-end against a real SketchUp-SDK-authored fixture.
// ---------------------------------------------------------------------------------------------

TEST(Edit, RealFixtureRoundTripsWithConsistentTopLevelCounts) {
  SkpModel original = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).parse();

  OpenExistingResult result = open_existing(test::fixture("capilla_quiroz_v17.skp"));
  ASSERT_NE(result.builder, nullptr);

  ByteBuffer rebuilt_bytes = result.builder->to_bytes();
  SkpModel rebuilt = SkpFile::from_buffer(rebuilt_bytes).parse();

  // Materials and layers are replayed 1:1 (every source material/layer becomes exactly one
  // writer call - see edit.hpp's own docstring for what's replayed vs approximated).
  EXPECT_EQ(rebuilt.materials.size(), original.materials.size());
  EXPECT_EQ(rebuilt.layers.size(), original.layers.size());

  // Every placed thing (group or instance alike) is replayed as a plain component instance
  // (documented fidelity gap), so root instance count should be at least as large as the
  // original's (never fewer, since nothing is silently dropped without a warning) - definitions
  // with no replayable geometry are skipped (and warned about), so definition count can be lower
  // but never higher.
  EXPECT_LE(rebuilt.definitions.size(), original.definitions.size());
}

}  // namespace
}  // namespace openskp
